app/url_utils.py

```diff
@@ -0,0 +1,24 @@
+from urllib.parse import urlsplit, urlunsplit
+
+
+def force_relative_url(base_url, url):
+    """
+    Forcibly removes the hostname and port from a URL, returning a new URL relative to base_url.
+    :param base_url: the path segment of base_url will be used to shorten the return url if possible
+    :param url: the URL to process into a relative URL
+    :return: sanitised URL relative to base_url
+    """
+    # convert args to str; urllib can accept bytes but our string ops can't
+    base_url, url = str(base_url), str(url)
+
+    # remove the netloc and port from the url...
+    parsed_url = urlsplit(url)
+    sanitised_url = urlunsplit(('', '', parsed_url.path, parsed_url.query, parsed_url.fragment))
+
+    # ...and remove any common path prefix up to and including the final slash
+    base_path = urlsplit(base_url).path or '/'
+    base_path = base_path[:base_path.rfind('/') + 1]
```

Comment:
Why do we care about the preceding slash? why not always have it or always don't have it?
Also why do we care about the matching bits? Can't we just store this as 
```
url = urlsplit(str(url))
urlunsplit(['', '', url.path, url.query, url.fragment])
```

Reply:
So it seemed weird to have a field where the first character was _always_ a slash.

So then I wanted a consistent/logical process for getting rid of that first slash, and I wanted something that could be expressed in terms of an explainable function. So this is really to cover some edge cases that shouldn't ever be seen in real life, but are (somewhat?) explained by the tests.

I'll admit that those lines only got added after I'd factored the code into a separate function and given it a name, and then thought about the implications of the edge cases. They aren't really essential to what we're doing, but making this a function that will behave consistently with its 'reverse' i.e. `urljoin` seemed like a nice idea.

Reply:
It only removes the preceding slash in some cases though. That's the bit that I don't get.

Reply:
Yup, this is quite hard to explain. The case where we don't do anything at all is where there isn't any kind of common prefix: this means that *if* there is some non-empty path in the base URL, that is not a prefix of the URL we are trying to shorten, then we are going to have to eliminate that by making sure there's a slash on the frond.

The "is not a prefix" thing in that previous sentence then needs clarification, because in general you don't have to consider anything after the final slash when deciding if something is a prefix, because of how relative paths are interpreted.

Anyway, this is what the last test case is trying to cover, and it shouldn't ever happen with the kinds of URL we will be passing around.

Reply:
but saying `c:my/weird/path` is the same as saying `c:/my/weird/path` so it doesn't matter if we always have a preceeding slash.
So we could just do this:
```
url = urlsplit(str(url))
path_string = urlunsplit(['', '', url.path, url.query, url.fragment])
return path_string if path_string.startswith('/') else '/' + path_string
```

Reply:
Grim. Tested this on my windows box at home
`c:my/weird/path is the same as saying c:/my/weird/path`
It's not. I stand corrected. I thought `c:` pointed to the mount point as an absolute.

![](https://giphy.com/gifs/funny-windows-x3aWCBvBaElYA.gif)

Here's what I'm unsure of:
OK so the point of this story is to remove the hostname and port to prevent people injecting malicious urls into our db?

What do you mean by 'mounted somewhere other than `/`'?

If all the paths are going to be relative then why does it matter if they begin with a slash or not?

Reply:
I _think_ George means in the case that we change our search-api to be mounted at, say, www.digitalmarketplace.service.gov.uk/search-api/